Add on premise microk8s deployment guide

fiedlerNr9 opened the pull request:
This PR adds a tutorial which showcases an on-premise deployment of Flyte using microk8s. 

The tutorial covers steps from setting up a on-premise device(raspberry pi) in a local network to submitting workflows from another development machine against the healthy Flyte deployment. To keep the overview on docs, a infra level directory structure was added as well. 

Open To Do's:
- With this directory structure, markdown links for the aws & k3s deployment needs to be adjusted

davidmirror-ops commented on docs/on-premises/microk8s/01-preparing-raspberry-pi.md:
```diff
@@ -0,0 +1,34 @@
+# Preparing the Raspberry Pi
+
+Feel free to skip this section, if you already have a ubuntu-server powered machine in your homelab. This section will be a brief walk through the process of preparing a Raspberry Pi before installing microk8s.
```
Ubuntu Server

davidmirror-ops commented on docs/on-premises/microk8s/01-preparing-raspberry-pi.md:
```diff
@@ -0,0 +1,34 @@
+
+## Pre-requisites
+### Hardware
+- Raspberry Pi 4 Modell B (8 GB RAM recommended)
```
```suggestion
- Raspberry Pi 4 Model B (8 GB RAM recommended)
```

davidmirror-ops commented on docs/on-premises/microk8s/01-preparing-raspberry-pi.md:
```diff
@@ -0,0 +1,34 @@
+
+## Flash the Micro SD Card
+- Download and run the Raspberry Pi Imager, choose your correct Raspberry Pi Device and select the Micro SD card that is connected to your machine
+- For the Operating System I recommend: `Other general-purpose OS > Ubuntu > Ubuntu Server 22.04.3 LTS (64-bit)`
```
```suggestion
- For the Operating System use: `Other general-purpose OS > Ubuntu > Ubuntu Server 22.04.3 LTS (64-bit)`
```

davidmirror-ops commented on docs/on-premises/microk8s/02-install-configure-microk8s.md:
```diff
@@ -0,0 +1,27 @@
+# Install and Configure MikroK8s
+
+To deploy kubernetes on your ubuntu powered device install the microk8s snap.
```
```suggestion
To deploy Kubernetes on your Ubuntu-powered device install the microk8s snap.
```

davidmirror-ops commented on docs/on-premises/microk8s/02-install-configure-microk8s.md:
````diff
@@ -0,0 +1,27 @@
+``` bash
+sudo snap install microk8s --classic --channel=1.25
+```
+To enable your current user seamless usage of microk8s command, run the following commands:
````
```suggestion
To enable the `microk8s` command for your current user,  run the following commands:
```

davidmirror-ops commented on docs/on-premises/microk8s/02-install-configure-microk8s.md:
````diff
@@ -0,0 +1,27 @@
+``` bash
+microk8s status
+```
+If this successfully displays the microk8s status, your are ready to operate via kubectl like shown in the following example:
````
```suggestion
If it successfully displays the microk8s status, you are ready to operate via `kubectl` like shown in the following example:
```

davidmirror-ops commented on docs/on-premises/microk8s/04-install-flyte.md:
```diff
@@ -0,0 +1,46 @@
+# Install Flyte
+To make sure your future Flyte deployment can access kubernetes services via internal dns, make sure to enable the dns addon of microk8s:
```
```suggestion
To make sure your future Flyte deployment can access Kubernetes services via internal DNS, make sure to enable the DNS addon of microk8s:
```

davidmirror-ops commented on docs/on-premises/microk8s/04-install-flyte.md:
````diff
@@ -0,0 +1,46 @@
+```
+> Make sure to adjust the values for the minio & postgres deployment you just submitted
+
+Install the Flyte binary via helm:
````
```suggestion
Install the Flyte binary via Helm:
```

davidmirror-ops commented on docs/on-premises/microk8s/05-add-ingress-and-tls.md:
```diff
@@ -0,0 +1,67 @@
+# Adding Ingress & TLS
+
+This section will demonstrate how to add ingress and tls to your flyte deployment in order to access flyte from any device in your local network.
```
```suggestion
This section will demonstrate how to add Ingress and TLS to your Flyte deployment to access Flyte from any device in your local network.
```

davidmirror-ops commented on docs/on-premises/microk8s/05-add-ingress-and-tls.md:
```diff
@@ -0,0 +1,67 @@
+
+This section will demonstrate how to add ingress and tls to your flyte deployment in order to access flyte from any device in your local network.
+
+First of all, enable the ingress microk8s addon:
```
```suggestion
First of all, enable the Ingress microk8s addon:
```

davidmirror-ops commented on docs/on-premises/microk8s/05-add-ingress-and-tls.md:
````diff
@@ -0,0 +1,67 @@
+microk8s enable ingress
+```
+
+> For simplicity reasons, this tutorial uses self-signed certificates to enable a tls connection.
````
```suggestion
> For simplicity reasons, this tutorial uses self-signed certificates to enable a TLS connection.
```

davidmirror-ops commented on docs/on-premises/microk8s/05-add-ingress-and-tls.md:
````diff
@@ -0,0 +1,67 @@
+openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout ${KEY_FILE} -out ${CERT_FILE} -subj "/CN=${HOST}/O=${HOST}" -addext "subjectAltName = DNS:${HOST}"
+```
+
+Given the certificate and key, create a kubernetes secret in the flyte namespace:
````
```suggestion
Given the certificate and key, create a Kubernetes secret in the `flyte` namespace:
```

davidmirror-ops commented on docs/on-premises/microk8s/05-add-ingress-and-tls.md:
````diff
@@ -0,0 +1,67 @@
+kubectl create secret tls ${CERT_NAME} --key ${KEY_FILE} --cert ${CERT_FILE} -n flyte
+```
+
+To add ingress and tls to your flyte deployment, add the following yaml to your Flyte binary helm chart and update your deployment:
````
```suggestion
To add Ingress and TLS to your Flyte deployment, add the following block to your Flyte binary Helm chart and update your deployment:
```

davidmirror-ops commented on docs/on-premises/microk8s/05-add-ingress-and-tls.md:
````diff
@@ -0,0 +1,67 @@
+  grpcAnnotations:
+    nginx.ingress.kubernetes.io/backend-protocol: GRPC
+```
+To validate your updated deployment, check the ingress resources in the flyte namespace:
````
```suggestion
To validate your updated deployment, check the Ingress resources in the flyte namespace:
```

fiedlerNr9 commented:
Thank you so much @davidmirror-ops! I incorporated your suggestions and updated some READMEs according to the new repo structure. I also went through all the steps again with a fresh raspberry and added some missing notes. For the future i might add chapters for auth and logs (microk8s observability addon) :)

Are you able to resolve the conflict on this PR? Not sure if i can do it with the permissions i have on this repo
